lib/Dialect/TTNN/Analysis/ShardSolver.cpp

```diff
@@ -520,10 +520,74 @@ bool ShardSolver::checkShardCompatible(
   //
 
   if (OpModel backend = dyn_cast<OpModel>(consumerOp)) {
-    if (false ==
-        backend.isOpLegal(std::vector{producerLayout}, consumerLayout)) {
+
+    auto deviceAttr = mlir::tt::getCurrentScopeDevice(producerOp);
+    assert(deviceAttr);
+    auto workerGrid = deviceAttr.getWorkerGrid();
+
+    // map consumer operands to DRAM interleave or provided producerLayout
+    // only one operand can be mapped to producerLayout, it's picked as first
+    // operand matching producerOp output shape.
+
+    uint32_t numOperands = consumerOp->getNumOperands();
+    // some ops have multiple operands; and some ops have output also an
+    // operand. TBD if there is a more robust way to get real number of inputs
+
+    // cast to DPSop?  //
+    numOperands = (numOperands > 1) ? numOperands - 1 : numOperands;
+    std::vector<TTNNLayoutAttr> inputLayouts;
+
+    auto inputUnderCheck =
+        mlir::cast<RankedTensorType>(producerOp->getResult(0).getType());
+    bool inputUnderCheckFound = false;
+
+    for (uint32_t i = 0; i < numOperands; i++) {
+      auto operand = consumerOp->getOperand(i);
+      auto input = mlir::cast<RankedTensorType>(operand.getType());
+
+      if ((inputUnderCheckFound == false) &&
+          (inputUnderCheck.getShape() == input.getShape())) {
+        // this is the input we are checking compatibility for
+        inputUnderCheckFound = true;
+        inputLayouts.push_back(producerLayout);
+      } else {
+        // this is the other input that we DRAM interleave
+
+        // what if it is tilized already?
+        auto elementType =
+            TileType::get(consumerOp->getContext(), input.getElementType());
+
+        auto layout = TTNNLayoutAttr::get(
+            consumerOp->getContext(), input.getShape(), elementType,
+            BufferType::DRAM, workerGrid,
+            TensorMemoryLayoutAttr::get(consumerOp->getContext(),
+                                        TensorMemoryLayout::Interleaved));
+        inputLayouts.push_back(layout);
+      }
+    }
```

Comment:
As discussed offline, a bit of cleanup is needed here. I'll do this in a follow up PR.